code-monkey-ts2/agents/agent.ts

```diff
@@ -0,0 +1,211 @@
+import fs from "fs";
+import path from "path";
+import { randomUUID } from "crypto";
+
+import { createReactAgent } from "@langchain/langgraph/prebuilt";
+import { MemorySaver } from "@langchain/langgraph";
+import { HumanMessage, SystemMessage } from "@langchain/core/messages";
+import { StructuredTool } from "@langchain/core/tools";
+
+import { instrument, currentSpan } from "../instrumentation";
+import { getRootDir, getArtifactsDir, getAgentMSN } from "../constants";
+import { MSN } from "../models";
+import { showDiff, askUser } from "../tools/utils";
+
+import { BaseAgent } from "./base_agent";
+
+export abstract class Agent extends BaseAgent {
+    model: ReturnType<typeof createReactAgent>;
```

Comment:
We always use 2 space-indent for everything where possible.
(I'm not sure why I was not able to get VSCode to use 2 space-indent for python, hence we ended up with 4 spaces there)